category/traversable/basic.lean

```diff
@@ -0,0 +1,105 @@
+/-
+Copyright (c) 2017 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Simon Hudon
+
+Type class for traversing collections
+-/
+
+import tactic.cache
+import category.applicative
+
+open function (hiding comp)
+
+universe variables u v w
+
+section applicative_transformation
+
+variables (f : Type u → Type v) [applicative f] [is_lawful_applicative f]
+variables (g : Type u → Type w) [applicative g] [is_lawful_applicative g]
+
+structure applicative_transformation : Type (max (u+1) v w) :=
+(F : ∀ {α : Type u}, f α → g α)
+(preserves_pure' : ∀ {α : Type u} (x : α), F (pure x) = pure x)
+(preserves_seq' : ∀ {α β : Type u} (x : f (α → β)) (y : f α), F (x <*> y) = F x <*> F y)
+
+instance : has_coe_to_fun (applicative_transformation f g) :=
+{ F := λ _, ∀ {α}, f α → g α,
+  coe := λ m, m.F }
+
+variables {f g}
+variables (η : applicative_transformation f g)
+
+@[functor_norm]
+lemma applicative_transformation.preserves_pure :
+  ∀ {α : Type u} (x : α), η (pure x) = pure x :=
+by apply applicative_transformation.preserves_pure'
+
+@[functor_norm]
+lemma applicative_transformation.preserves_seq :
+  ∀ {α β : Type u} (x : f (α → β)) (y : f α), η (x <*> y) = η x <*> η y :=
+by apply applicative_transformation.preserves_seq'
+
+@[functor_norm]
+lemma applicative_transformation.preserves_map {α β : Type u} (x : α → β)  (y : f α) :
+  η (x <$> y) = x <$> η y :=
+by rw [← pure_seq_eq_map,η.preserves_seq]; simp with functor_norm
+
+end applicative_transformation
+
+class traversable (t : Type u → Type u)
+extends functor t :=
```

Comment:
`extends` in the same line as `class` (if it doesn't fit in one line at least indent it)